app/multitenant/consul_pipe_router.go

```diff
@@ -0,0 +1,431 @@
+		select {
+		case <-pr.quit:
+			return false
+		default:
+		}
+
+		pr.actorChan <- func() { pr.handlePipeUpdate(key, value.(*consulPipe)) }
+		return true
+	})
+}
+
+func (pr *consulPipeRouter) handlePipeUpdate(key string, cp *consulPipe) {
+	log.Infof("Got update to pipe %s", key)
+
+	// 1. If this pipe is closed, or we're not one of the ends, we
+	//    should ensure our local pipe (and bridge) is closed.
+	if !cp.DeletedAt.IsZero() || !cp.eitherEndFor(pr.advertise) {
+		log.Infof("Pipe %s not in use on this node.", key)
+		pipe, ok := pr.activePipes[key]
+		delete(pr.activePipes, key)
+		if ok {
+			pipe.Close()
+		}
+
+		bridge, ok := pr.bridges[key]
+		delete(pr.bridges, key)
+		if ok {
+			bridge.stop()
+		}
+		return
+	}
+
+	if !cp.eitherEndFor(pr.advertise) {
+		return
+	}
+
+	// 2. If this pipe if for us, we should have a pipe for it.
+	pipe, ok := pr.activePipes[key]
+	if !ok {
+		pipe = xfer.NewPipe()
+		pr.activePipes[key] = pipe
+	}
+
+	// 3. Ensure there is a bridging connection for this pipe.
+	//    Semantics are the owner of the UIEnd connects to the owner of the ProbeEnd
+	shouldBridge := cp.DeletedAt.IsZero() &&
+		cp.addrFor(app.UIEnd) != cp.addrFor(app.ProbeEnd) &&
+		cp.addrFor(app.UIEnd) == pr.advertise &&
+		cp.addrFor(app.ProbeEnd) != ""
+	bridge, ok := pr.bridges[key]
+
+	// If we shouldn't be bridging but are, or we should be bridging but are pointing
+	// at the wrong place, stop the current bridge.
+	if (!shouldBridge && ok) || (shouldBridge && ok && bridge.addr != cp.addrFor(app.ProbeEnd)) {
+		log.Infof("Stopping bridge connection for %s", key)
+		delete(pr.bridges, key)
+		bridge.stop()
+		ok = false
+	}
+
+	// If we should be bridging and are not, start a new bridge
+	if shouldBridge && !ok {
+		log.Infof("Starting bridge connection for %s", key)
+		bridge = newBridgeConnection(key, cp.addrFor(app.ProbeEnd), pipe)
+		pr.bridges[key] = bridge
+	}
+}
+
+func (pr *consulPipeRouter) privateAPI() {
+	router := mux.NewRouter()
+	router.Methods("GET").
+		MatcherFunc(app.URLMatcher("/private/api/pipe/{key}")).
+		HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			var (
+				key = mux.Vars(r)["key"]
+				pc  = make(chan xfer.Pipe)
+			)
+			pr.actorChan <- func() {
+				pc <- pr.activePipes[key]
+			}
+			pipe := <-pc
+			if pipe == nil {
+				http.NotFound(w, r)
+				return
+			}
+
+			conn, err := xfer.Upgrade(w, r, nil)
+			if err != nil {
+				log.Errorf("Error upgrading pipe %s websocket: %v", key, err)
+				return
+			}
+			defer conn.Close()
+
+			end, _ := pipe.Ends()
+			if err := pipe.CopyToWebsocket(end, conn); err != nil && !xfer.IsExpectedWSCloseError(err) {
+				log.Printf("Error copying to pipe %s websocket: %v", key, err)
+			}
+		})
+
+	addr := fmt.Sprintf("%s:%d", pr.advertise, privateAPIPort)
+	log.Infof("Serving private API on endpoint %s.", addr)
+	log.Infof("Private API terminated: %v", http.ListenAndServe(addr, router))
```

Comment:
Yes, but we've never cared about it in the past in places like prog/app.go (these things all existing for the length of the process)